modules/core/src/main/java/org/apache/ignite/internal/event/AbstractEventProducer.java

```diff
@@ -81,20 +81,32 @@ protected CompletableFuture<Void> fireEvent(T evt, P params, @Nullable Throwable
             return nullCompletedFuture();
         }
 
-        CompletableFuture<?>[] futures = new CompletableFuture[listeners.size()];
+        // Lazy init.
+        List<CompletableFuture<?>> futures = null;
 
         for (int i = 0; i < listeners.size(); i++) {
             EventListener<P> listener = listeners.get(i);
 
-            futures[i] = listener.notify(params, err)
-                    .thenAccept(remove -> {
-                        if (remove) {
-                            removeListener(evt, listener);
-                        }
-                    });
+            CompletableFuture<Boolean> future = listener.notify(params, err);
+
+            if (future.isDone() && !future.isCompletedExceptionally()) {
```

Comment:
Shall we also check that the future is not canceled?

Reply:
No, `isCompletedExceptionally` covers that